CONFIG.md
## Config
All aspects of the blood moon can be configured, below is an example of a config file which would enable all of the features. More details on each section can be found below. Each section contains an enabled setting, this controls if that feature will be used or not. You will almost certainly want to configure the plugin to your needs, by default everything is enabled. One of these files will be created for each world that exists on the server, there is no global config.

```
enabled: false
always-on: false
chance: 14
features:
  chat-message:
    enabled: true
    message: '&cThe bloodmoon is rising !'
  server-commands:
    enabled: false
    commands:
      start:
      - toggledownfall
      - time set 0
      - op Notch
      end:
      - toggledownfall
      - time set 12000
      - deop Notch
  play-sound:
    enabled: true
    sound: WITHER_SPAWN
    pitch: 1.0
    volume: 1.0
  arrow-rate:
    enabled: true
    multiplier: 2
  fire-arrows:
    enabled: true
    chance: 100
    ignite-target: true
  zombie-weapon:
    enabled: true
    chance: 60
    drop-chance: 25
    ignore-spawners: true
    weapons:
    - DIAMOND_SWORD
    - GOLD_SWORD
    - IRON_SWORD
  zombie-armor:
    enabled: true
    chance: 60
    drop-chance: 7
    ignore-spawners: true
    armor:
    - DIAMOND
    - GOLD
    - IRON
  target-distance:
    enabled: true
    multiplier: 3
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
  movement-speed:
    enabled: true
    multiplier: 1.3
    fast-chance: 15
    fast-multiplier: 1.5
    mobs:
    - ZOMBIE
    - SKELETON
    - CREEPER
  break-blocks:
    enabled: true
    drop-items: true
    realistic-drop: true
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    blocks:
    - WOOD
    - LOG
    - GLASS
  max-health:
    enabled: true
    multiplier: 2.0
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
  more-spawning:
    enabled: true
    multiplier: 2
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
  more-exp:
    enabled: true
    ignore-spawners: false
    multiplier: 2
  more-drops:
    enabled: true
    ignore-spawners: false
    multiplier: 2
  sword-damage:
    enabled: true
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    chance: 10
  super-creepers:
    enabled: true
    power: 4.0
    fire: true
  spawn-on-kill:
    enabled: true
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    chance: 10
    spawn:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
  spawn-on-sleep:
    enabled: true
    spawn:
    - ZOMBIE
  spawn-control:
    enabled: true
    spawn:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    - PIG_ZOMBIE
    - BLAZE
    - MAGMA_CUBE
  lock-in-world:
    enabled: false
  texture-pack:
    enabled: false
    normal: http://bukkit.jacekk.co.uk/bloodmoon_tps/normal.zip
    bloodmoon: http://bukkit.jacekk.co.uk/bloodmoon_tps/bloodmoon.zip
  extended-night:
    enabled: true
    min-kills: 16
    max-resets: 6
  weather:
    enabled: true
    thunder: true
    rain: true
  daylight-proof-mobs:
    enabled: true
  nether-sky:
    enabled: false
  dungeons:
    enabled: true
    biomes:
    - PLAINS
    - ICE_PLAINS
    - DESERT
    - SWAMPLAND
    chance: 10
    min-layers: 3
    max-layers: 5
    chest-items:
    - BREAD
    - APPLE
    - PORK
    - SADDLE
    - BUCKET
    - STRING
    - REDSTONE
    - SULPHUR
    - COCOA
    min-stack-size: 1
    max-stack-size: 8
    items-per-chest: 12
    spawner-types:
    - ZOMBIE
    - SKELETON
    spawner-delay: 100
    spawner-count: 6
    spawner-max-mobs: 8
```

## Main Options
```
enabled: false
always-on: false
chance: 14
```
Setting enabled to true will enable the plugin for this world, when disabled none of the other options have any effect. Setting always-on to true will override the chance and force the blood moon to be active all of the time. The chance setting controls how often a blood moon will occur, this is a percentage so setting it to 100 will cause the blood moon to happen every night.

## Features
### chat-message
```
  chat-message:
    enabled: true
    message: '&cThe bloodmoon is rising !'
```
This section controls the chat message that is sent to players when they enter a world that has a bloodmoon or one starts in their world. The standard colour codes can be used.

### server-commands
```
  server-commands:
    enabled: false
    commands:
      start:
      - toggledownfall
      - time set 0
      - op Notch
      end:
      - toggledownfall
      - time set 12000
      - deop Notch
```
This section controls server commands that are executed when a bloodmoon starts and ends, allowing for customisation with other plugins and server features. The commands.start and commands.end sections list the commands that are executed if this feature is enabled.

### play-sound
```
  play-sound:
    enabled: true
    sound: WITHER_SPAWN
    pitch: 1.0
    volume: 1.0
```
If this feature is enabled a sound will be played when the blood moon starts. The sound option controls which sound is played and the pitch and volume options modify the pitch and volume of the sound. For a full list of supported sounds see the Bukkit JavaDoc

### arrow-rate
```
  arrow-rate:
    enabled: true
    multiplier: 2
```
This section controls the rate at which skeletons will shoot arrows during a bloodmoon the multiplier option controls how fast arrows will be fired.

**WARNING:** Setting multiplier to a value above 20 will cause a divide by zero error.

### fire-arrows
```
  fire-arrows:
    enabled: true
    chance: 100
    ignite-arget: true
```
This section controls if the arrows that skeletons shoot should be set alight or not, the ignite-target option controls if they should set the blocks they hit alight too.

### zombie-weapon
```
  zombie-weapon:
    enabled: true
    chance: 60
    drop-chance: 25
    weapons:
    - DIAMOND_SWORD
    - GOLD_SWORD
    - IRON_SWORD
```
This section controls if zombies should spawn with weapons in their hand, the chance is how likely it is that a zombie will have a weapon and drop-chance is how likely they are to drop it when they are killed. The name of any item can be used in the weapons list, an entry is picked from here at random.

### zombie-armor
```
  zombie-armor:
    enabled: true
    chance: 60
    drop-chance: 7
    armor:
    - DIAMOND
    - GOLD
    - IRON
```
This section controls if zombies should spawn wearing armor, the chance is how likely it is that a zombie will have armor and drop-chance is how likely they are to drop a single part of it when they are killed. The type of armor is picked at random from the armor list.

### target-distance
```
  target-distance:
    enabled: true
    multiplier: 3
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
```
This section controls how much further mobs will target players from, the mobs list controls which mobs this will affect. Setting the multiplier to a high value will cause a significant drop in performance.

### movement-speed
```
  movement-speed:
    enabled: true
    multiplier: 1.3
    fast-chance: 15
    fast-multiplier: 1.5
    mobs:
    - ZOMBIE
    - SKELETON
    - CREEPER
```
This section controls how much faster mobs should move. Setting the multiplier option to 1.0 will leave the mobs at their normal speed, the default of 1.35 make zombies move slightly faster than spiders do by default. fast-chance is the percentage chance that the mob will move at the fast-multiplier instead of the previous one.

### break-blocks
```
  break-blocks:
    enabled: true
    drop-items: true
    realistic-drop: true
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    blocks:
    - WOOD
    - LOG
    - GLASS
```
This section allows you to configure which mobs are able to break the blocks they are facing as well as which blocks can be broken. The mobs section is the same as for the above double-health section except that it controls which mobs can break the blocks. If the realistic-drop option is set to true the blocks will drop the same items as they would if broken by a player, if not (the default) they will drop their block type. For example, a glowstone block will drop glowstone dust with realistic-drop set to true and a glowstone block with it set to false, if stop-items is set to false no items will drop at all (useful for large servers). Finally the blocks section controls which blocks can be broken, their names should be used and as with the mob lists they must be in upper-case.

### max-health
```
  double-health:
    enabled: true
    multiplier: 2.0
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
```
This section controls which mobs should have their health increased, the mobs setting should contain a list of mob names to apply this feature to and the multiplier option controls how much higher the health should be.

### more-spawning
```
  more-spawning:
    enabled: true
    multiplier: 2
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
```
This section allows you to control how many more mobs than usual should spawn, the multiplier controls how many more should spawn. The mobs list controls which mobs this should affect.

### more-exp
```
  more-exp:
    enabled: true
    ignore-spawners: false
    multiplier: 2
```
This section controls how much more experience orbs should be dropped when a player kills a mob. The ignore-spawners option makes this not affect mobs that came form spawners and the multiplier controls how much more XP should be dropped.

### more-drops
```
  more-drops:
    enabled: true
    ignore-spawners: false
    multiplier: 2
```
This section controls how many more items should be dropped when a player kills a mob. The ignore-spawners option makes this not affect mobs that came form spawners and the multiplier controls how many more items should be dropped.

### sword-damage
```
  sword-damage:
    enabled: true
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    chance: 10
```
This section allows you to configure the random large damage that can be done to weapons when attacking mobs. The mobs section controls which mobs this should work for, as above this needs to be a list of their names in upper-case. The chance option controls the percentage chance that a weapon will receive large damage with every strike of one of the mobs on the list.

### super-creepers
```
  super-creepers:
    enabled: true
    power: 4.0
    fire: true
```
This section controls the modification of creeper explosions, with this enabled the explosions will cause fires to start on the destroyed blocks. The power setting controls the size of the explosion, the default of 4.0 is the same as a TNT explosion, there is no limit imposed on this value so be sensible. The fire setting controls if the explosions should start fires or not.

### spawn-on-kill
```
  spawn-on-kill:
    enabled: true
    mobs:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    chance: 10
    spawn:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
```
This controls if a mob should be spawned when a player kills a mob. The spawn section is the same as for spawn-on-sleep it should contains a list of mobs that might spawn. The mobs section controls which mobs this applies to, and the chance option controls the percentage chance that a mob will spawn. There is no limit imposed on the chance but setting it to anything close to 100 will be very irritating.

### spawn-on-sleep
```
  spawn-on-sleep:
    enabled: true
    spawn:
    - ZOMBIE
```
This controls if a mob should be spawned if a player tries to sleep during a blood moon. The spawn section should specify a list of mobs, one mob from this list will be randomly picked and spawned if a player tries to use a bed.

### spawn-control
```
  spawn-control:
    enabled: true
    spawn:
    - ZOMBIE
    - SKELETON
    - SPIDER
    - CREEPER
    - ENDERMAN
    - PIG_ZOMBIE
    - BLAZE
    - MAGMA_CUBE
```
This section allows you to modify the mobs that will spawn during a bloodmoon, only the entities names in the spawn list will appear. You can add mobs that do not naturally spawn in the overworld.

### lock-in-world
```
  lock-in-world:
    enabled: true
```
If this is set to false the player will not be allowed to leave the world they are in until the bloodmoon has ended. New players will still be able to enter so this may be seen as a little unfair.

### texture-pack
```
  texture-pack:
    enabled: false
    normal: http://bukkit.jacekk.co.uk/bloodmoon_tps/normal.zip
    bloodmoon: http://bukkit.jacekk.co.uk/bloodmoon_tps/bloodmoon.zip
```
These texture packs are sent to the client depending on the bloodmoon state, the default values for this feature simply make the moon red during a bloodmoon.

### extended-night
```
  extended-night:
    enabled: true
    min-kills: 16
    max-resets: 6
```
This feature will cause the night to repeat if the required number of mobs have not been killed. The night will stop repeating after mine-kills mobs have been killed or it has already reset max-resets times.

### weather
```
  weather:
    enabled: true
    thunder: true
    rain: true
```
This section allows you to control the weather during a blood moon, setting rain to true will make it rain until the event is over and thunder will make it thunder until the event is over.

### daylight-proof-mobs
```
  daylight-proof-mobs:
    enabled: true
```
If enabled mobs will not take damage from being in sunlight, useful for always-on mode.

### nether-sky
```
  nether-sky:
    enabled: false
```
If this feature is enabled players will see the sky from the nether in all worlds with an active bloodmoon, this defaults to false as it causes the player to see the respawn screen when the bloodmoon starts.

### dungeons
```
  dungeons:
    enabled: true
    biomes:
    - PLAINS
    - ICE_PLAINS
    - DESERT
    - SWAMPLAND
    chance: 10
    min-layers: 3
    max-layers: 5
    chest-items:
    - BREAD
    - APPLE
    - PORK
    - SADDLE
    - BUCKET
    - STRING
    - REDSTONE
    - SULPHUR
    - COCOA
    min-stack-size: 1
    max-stack-size: 8
    items-per-chest: 12
    spawner-types:
    - ZOMBIE
    - SKELETON
    spawner-delay: 100
    spawner-count: 6
    spawner-max-mobs: 8
```
This section controls the occurrence and features of the dungeons that can be generated in the world. These are much larger than the vanilla ones and will typically be much more hostile. They are also the only feature which is always present in the world regardless of the time.

- biomes - A list of biomes which will generate dungeons, they work much better in flat areas and can cause problems in jungles where there are lots of other structures.
- chance - The chance that each chunk has of containing a dungeon. this is a fraction of 10,000
- min-layers - The minimum number of floors that a dungeon can have
- max-layers - The maximum number of floors that a dungeon can have, note that setting this too high will result in them reaching the bottom of the would and there being no loot being added.
- chest-items - A list of items which can be found in the loot chests.
- min-stack-size - The minimum number of items which will be placed into each stack in the chest.
- max-stack-size - The maximum number of items which will be placed into each stack in the chest. Items which cannot normally be stacked will only ever generate as a single item.
- items-per-chest - The number of items to place in each of the four chests in the loot room. This is actually the number of times an item will be placed into a random slot in the chest meaning that less items can be found if the same slot if picked more than once.
- spawner-types - The mob types to place spawners for.
- spawner-delay - The time in ticks between spawn attempts from a spawner, a lower value will make it harder to break a spawner without more mobs spawning. The vanilla dungeons use a value between 600 and 800 for this.
- spawner-count - The number of attempts a spawner will make to spawn a mob after each delay, a higher value will mean that players become overrun much faster. Vanilla dungeons use a value of 4.
- spanwer-max-mobs - The number of mobs that need to be around a spawner to prevent it from spawning, each floor of the dungeon has 2 spawners so this will roughly be half of the number of mobs found on each floor. Vanilla dungeons use a value of 4.


src/main/java/uk/co/jacekk/bukkit/bloodmoon/Config.java
package uk.co.jacekk.bukkit.bloodmoon;

import org.bukkit.Material;
import org.bukkit.block.Biome;
import org.bukkit.entity.EntityType;
import uk.co.jacekk.bukkit.baseplugin.config.PluginConfigKey;

import java.util.Arrays;

public class Config {

    public static final PluginConfigKey ENABLED = new PluginConfigKey("enabled", false);
    public static final PluginConfigKey ALWAYS_ON = new PluginConfigKey("always-on", false);
    public static final PluginConfigKey CHANCE = new PluginConfigKey("chance", 14);

    public static final PluginConfigKey FEATURE_CHAT_MESSAGE_ENABLED = new PluginConfigKey("features.chat-message.enabled", true);
    public static final PluginConfigKey FEATURE_CHAT_MESSAGE_START_MESSAGE = new PluginConfigKey("features.chat-message.start-message", "&cThe bloodmoon is rising !");
    public static final PluginConfigKey FEATURE_CHAT_MESSAGE_END_MESSAGE = new PluginConfigKey("features.chat-message.end-message", "&cThe bloodmoon is over !");

    public static final PluginConfigKey FEATURE_SERVER_COMMANDS_ENABLED = new PluginConfigKey("features.server-commands.enabled", false);
    public static final PluginConfigKey FEATURE_SERVER_COMMANDS_START_COMMANDS = new PluginConfigKey("features.server-commands.commands.start", Arrays.asList("toggledownfall", "time set 0", "op Notch"));
    public static final PluginConfigKey FEATURE_SERVER_COMMANDS_END_COMMANDS = new PluginConfigKey("features.server-commands.commands.end", Arrays.asList("toggledownfall", "time set 12000", "deop Notch"));

    public static final PluginConfigKey FEATURE_DISABLED_COMMANDS_ENABLED = new PluginConfigKey("features.disabled-commands.enabled", true);
    public static final PluginConfigKey FEATURE_DISABLED_COMMANDS_COMMANDS = new PluginConfigKey("features.disabled-commands.commands", Arrays.asList("spawn", "home"));

    public static final PluginConfigKey FEATURE_PLAY_SOUND_ENABLED = new PluginConfigKey("features.play-sound.enabled", true);
    public static final PluginConfigKey FEATURE_PLAY_SOUND_SOUND = new PluginConfigKey("features.play-sound.sound", "ENTITY_WITHER_SPAWN");
    public static final PluginConfigKey FEATURE_PLAY_SOUND_PITCH = new PluginConfigKey("features.play-sound.pitch", 1.0d);
    public static final PluginConfigKey FEATURE_PLAY_SOUND_VOLUME = new PluginConfigKey("features.play-sound.volume", 1.0d);

    public static final PluginConfigKey FEATURE_ARROW_RATE_ENABLED = new PluginConfigKey("features.arrow-rate.enabled", true);
    public static final PluginConfigKey FEATURE_ARROW_RATE_MULTIPLIER = new PluginConfigKey("features.arrow-rate.multiplier", 2);

    public static final PluginConfigKey FEATURE_FIRE_ARROWS_ENABLED = new PluginConfigKey("features.fire-arrows.enabled", true);
    public static final PluginConfigKey FEATURE_FIRE_ARROWS_CHANCE = new PluginConfigKey("features.fire-arrows.chance", 100);
    public static final PluginConfigKey FEATURE_FIRE_ARROWS_IGNITE_TARGET = new PluginConfigKey("features.fire-arrows.ignite-target", true);

    public static final PluginConfigKey FEATURE_ZOMBIE_WEAPON_ENABLED = new PluginConfigKey("features.zombie-weapon.enabled", true);
    public static final PluginConfigKey FEATURE_ZOMBIE_WEAPON_CHANCE = new PluginConfigKey("features.zombie-weapon.chance", 60);
    public static final PluginConfigKey FEATURE_ZOMBIE_WEAPON_DROP_CHANCE = new PluginConfigKey("features.zombie-weapon.drop-chance", 25);
    public static final PluginConfigKey FEATURE_ZOMBIE_WEAPON_IGNORE_SPAWNERS = new PluginConfigKey("features.zombie-weapon.ignore-spawners", true);
    public static final PluginConfigKey FEATURE_ZOMBIE_WEAPON_WEAPONS = new PluginConfigKey("features.zombie-weapon.weapons", Arrays.asList("DIAMOND_SWORD", "GOLD_SWORD", "IRON_SWORD"));

    public static final PluginConfigKey FEATURE_ZOMBIE_ARMOR_ENABLED = new PluginConfigKey("features.zombie-armor.enabled", true);
    public static final PluginConfigKey FEATURE_ZOMBIE_ARMOR_CHANCE = new PluginConfigKey("features.zombie-armor.chance", 60);
    public static final PluginConfigKey FEATURE_ZOMBIE_ARMOR_DROP_CHANCE = new PluginConfigKey("features.zombie-armor.drop-chance", 7);
    public static final PluginConfigKey FEATURE_ZOMBIE_ARMOR_IGNORE_SPAWNERS = new PluginConfigKey("features.zombie-armor.ignore-spawners", true);
    public static final PluginConfigKey FEATURE_ZOMBIE_ARMOR_ARMOR = new PluginConfigKey("features.zombie-armor.armor", Arrays.asList("DIAMOND", "GOLD", "IRON"));

    public static final PluginConfigKey FEATURE_TARGET_DISTANCE_ENABLED = new PluginConfigKey("features.target-distance.enabled", true);
    public static final PluginConfigKey FEATURE_TARGET_DISTANCE_MULTIPLIER = new PluginConfigKey("features.target-distance.multiplier", 2.0d);
    public static final PluginConfigKey FEATURE_TARGET_DISTANCE_MOBS = new PluginConfigKey("features.target-distance.mobs", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER"));

    public static final PluginConfigKey FEATURE_MOVEMENT_SPEED_ENABLED = new PluginConfigKey("features.movement-speed.enabled", true);
    public static final PluginConfigKey FEATURE_MOVEMENT_SPEED_MULTIPLIER = new PluginConfigKey("features.movement-speed.multiplier", 1.20d);
    public static final PluginConfigKey FEATURE_MOVEMENT_SPEED_FAST_CHANCE = new PluginConfigKey("features.movement-speed.fast-chance", 15);
    public static final PluginConfigKey FEATURE_MOVEMENT_SPEED_FAST_MULTIPLIER = new PluginConfigKey("features.movement-speed.fast-multiplier", 1.3d);
    public static final PluginConfigKey FEATURE_MOVEMENT_SPEED_MOBS = new PluginConfigKey("features.movement-speed.mobs", Arrays.asList("ZOMBIE", "SKELETON", "CREEPER"));

    public static final PluginConfigKey FEATURE_BREAK_BLOCKS_ENABLED = new PluginConfigKey("features.break-blocks.enabled", true);
    public static final PluginConfigKey FEATURE_BREAK_BLOCKS_DROP_ITEMS = new PluginConfigKey("features.break-blocks.drop-items", true);
    public static final PluginConfigKey FEATURE_BREAK_BLOCKS_REALISTIC_DROP = new PluginConfigKey("features.break-blocks.realistic-drop", true);
    public static final PluginConfigKey FEATURE_BREAK_BLOCKS_MOBS = new PluginConfigKey("features.break-blocks.mobs", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN"));
    public static final PluginConfigKey FEATURE_BREAK_BLOCKS_BLOCKS = new PluginConfigKey("features.break-blocks.blocks", Arrays.asList("WOOD", "LOG", "GLASS"));

    public static final PluginConfigKey FEATURE_MAX_HEALTH_ENABLED = new PluginConfigKey("features.max-health.enabled", true);
    public static final PluginConfigKey FEATURE_MAX_HEALTH_MULTIPLIER = new PluginConfigKey("features.max-health.multiplier", 2.0d);
    public static final PluginConfigKey FEATURE_MAX_HEALTH_MOBS = new PluginConfigKey("features.max-health.mobs", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN"));

    public static final PluginConfigKey FEATURE_MORE_SPAWNING_ENABLED = new PluginConfigKey("features.more-spawning.enabled", true);
    public static final PluginConfigKey FEATURE_MORE_SPAWNING_MULTIPLIER = new PluginConfigKey("features.more-spawning.multiplier", 2);
    public static final PluginConfigKey FEATURE_MORE_SPAWNING_MOBS = new PluginConfigKey("features.more-spawning.mobs", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN"));

    public static final PluginConfigKey FEATURE_MORE_EXP_ENABLED = new PluginConfigKey("features.more-exp.enabled", true);
    public static final PluginConfigKey FEATURE_MORE_EXP_IGNORE_SPAWNERS = new PluginConfigKey("features.more-exp.ignore-spawners", false);
    public static final PluginConfigKey FEATURE_MORE_EXP_MULTIPLIER = new PluginConfigKey("features.more-exp.multiplier", 2);

    public static final PluginConfigKey FEATURE_MORE_DROPS_ENABLED = new PluginConfigKey("features.more-drops.enabled", true);
    public static final PluginConfigKey FEATURE_MORE_DROPS_IGNORE_SPAWNERS = new PluginConfigKey("features.more-drops.ignore-spawners", false);
    public static final PluginConfigKey FEATURE_MORE_DROPS_MULTIPLIER = new PluginConfigKey("features.more-drops.multiplier", 2);

    public static final PluginConfigKey FEATURE_SWORD_DAMAGE_ENABLED = new PluginConfigKey("features.sword-damage.enabled", true);
    public static final PluginConfigKey FEATURE_SWORD_DAMAGE_MOBS = new PluginConfigKey("features.sword-damage.mobs", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN"));
    public static final PluginConfigKey FEATURE_SWORD_DAMAGE_CHANCE = new PluginConfigKey("features.sword-damage.chance", 10);

    public static final PluginConfigKey FEATURE_SUPER_CREEPERS_ENABLED = new PluginConfigKey("features.super-creepers.enabled", true);
    public static final PluginConfigKey FEATURE_SUPER_CREEPERS_POWER = new PluginConfigKey("features.super-creepers.power", 4.0D);
    public static final PluginConfigKey FEATURE_SUPER_CREEPERS_FIRE = new PluginConfigKey("features.super-creepers.fire", true);
    public static final PluginConfigKey FEATURE_SUPER_CREEPERS_LIGHTNING = new PluginConfigKey("features.super-creepers.lightning", true);

    public static final PluginConfigKey FEATURE_SPAWN_ON_KILL_ENABLED = new PluginConfigKey("features.spawn-on-kill.enabled", true);
    public static final PluginConfigKey FEATURE_SPAWN_ON_KILL_MOBS = new PluginConfigKey("features.spawn-on-kill.mobs", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN"));
    public static final PluginConfigKey FEATURE_SPAWN_ON_KILL_CHANCE = new PluginConfigKey("features.spawn-on-kill.chance", 10);
    public static final PluginConfigKey FEATURE_SPAWN_ON_KILL_SPAWN = new PluginConfigKey("features.spawn-on-kill.spawn", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN"));

    public static final PluginConfigKey FEATURE_SPAWN_ON_SLEEP_ENABLED = new PluginConfigKey("features.spawn-on-sleep.enabled", true);
    public static final PluginConfigKey FEATURE_SPAWN_ON_SLEEP_SPAWN = new PluginConfigKey("features.spawn-on-sleep.spawn", Arrays.asList("ZOMBIE"));

    public static final PluginConfigKey FEATURE_SPAWN_CONTROL_ENABLED = new PluginConfigKey("features.spawn-control.enabled", true);
    public static final PluginConfigKey FEATURE_SPAWN_CONTROL_SPAWN = new PluginConfigKey("features.spawn-control.spawn", Arrays.asList("ZOMBIE", "SKELETON", "SPIDER", "CREEPER", "ENDERMAN", "PIG_ZOMBIE", "BLAZE", "MAGMA_CUBE"));

    public static final PluginConfigKey FEATURE_LOCK_IN_WORLD_ENABLED = new PluginConfigKey("features.lock-in-world.enabled", false);

    public static final PluginConfigKey FEATURE_TEXTURE_PACK_ENABLED = new PluginConfigKey("features.texture-pack.enabled", false);
    public static final PluginConfigKey FEATURE_TEXTURE_PACK_NORMAL = new PluginConfigKey("features.texture-pack.normal", "http://bukkit.jacekk.co.uk/bloodmoon_tps/normal.zip");
    public static final PluginConfigKey FEATURE_TEXTURE_PACK_BLOODMOON = new PluginConfigKey("features.texture-pack.bloodmoon", "http://bukkit.jacekk.co.uk/bloodmoon_tps/bloodmoon.zip");

    public static final PluginConfigKey FEATURE_EXTENDED_NIGHT_ENABLED = new PluginConfigKey("features.extended-night.enabled", true);
    public static final PluginConfigKey FEATURE_EXTENDED_NIGHT_MIN_KILLS = new PluginConfigKey("features.extended-night.min-kills", 16);

    public static final PluginConfigKey FEATURE_WEATHER_ENABLED = new PluginConfigKey("features.weather.enabled", true);
    public static final PluginConfigKey FEATURE_WEATHER_THUNDER = new PluginConfigKey("features.weather.thunder", true);
    public static final PluginConfigKey FEATURE_WEATHER_RAIN = new PluginConfigKey("features.weather.rain", true);
    public static final PluginConfigKey FEATURE_WEATHER_CHANCE = new PluginConfigKey("features.weather.chance", 50);

    public static final PluginConfigKey FEATURE_DAYLIGHT_PROOF_MOBS_ENABLED = new PluginConfigKey("features.daylight-proof-mobs.enabled", true);

    public static final PluginConfigKey FEATURE_NETHER_SKY_ENABLED = new PluginConfigKey("features.nether-sky.enabled", false);

    public static final PluginConfigKey FEATURE_DUNGEONS_ENABLED = new PluginConfigKey("features.dungeons.enabled", true);
    public static final PluginConfigKey FEATURE_DUNGEONS_PROTECTED = new PluginConfigKey("features.dungeons.protected", true);
    public static final PluginConfigKey FEATURE_DUNGEONS_BIOMES = new PluginConfigKey("features.dungeons.biomes", Arrays.asList(Biome.PLAINS.name(), Biome.ICE_FLATS.name(), Biome.DESERT.name(), Biome.SWAMPLAND.name()));
    public static final PluginConfigKey FEATURE_DUNGEONS_CHANCE = new PluginConfigKey("features.dungeons.chance", 10);
    public static final PluginConfigKey FEATURE_DUNGEONS_MIN_LAYERS = new PluginConfigKey("features.dungeons.min-layers", 3);
    public static final PluginConfigKey FEATURE_DUNGEONS_MAX_LAYERS = new PluginConfigKey("features.dungeons.max-layers", 5);
    public static final PluginConfigKey FEATURE_DUNGEONS_SPAWNER_TYPES = new PluginConfigKey("features.dungeons.spawner-types", Arrays.asList(EntityType.ZOMBIE.name(), EntityType.SKELETON.name()));
    public static final PluginConfigKey FEATURE_DUNGEONS_SPAWNER_DELAY = new PluginConfigKey("features.dungeons.spawner-delay", 100);
    public static final PluginConfigKey FEATURE_DUNGEONS_SPAWNER_COUNT = new PluginConfigKey("features.dungeons.spawner-count", 6);
    public static final PluginConfigKey FEATURE_DUNGEONS_SPAWNER_MAX_MOBS = new PluginConfigKey("features.dungeons.spawner-max-mobs", 8);
    public static final PluginConfigKey FEATURE_DUNGEONS_CHEST_ITEMS = new PluginConfigKey("features.dungeons.chest-items", Arrays.asList(Material.BREAD.name(), Material.APPLE.name(), Material.PORK.name(), Material.SADDLE.name(), Material.BUCKET.name(), Material.STRING.name(), Material.REDSTONE.name(), Material.SULPHUR.name(), Material.COCOA.name()));
    public static final PluginConfigKey FEATURE_DUNGEONS_MIN_STACK_SIZE = new PluginConfigKey("features.dungeons.min-stack-size", 1);
    public static final PluginConfigKey FEATURE_DUNGEONS_MAX_STACK_SIZE = new PluginConfigKey("features.dungeons.max-stack-size", 8);
    public static final PluginConfigKey FEATURE_DUNGEONS_ITEMS_PER_CHEST = new PluginConfigKey("features.dungeons.items-per-chest", 12);

    public static final PluginConfigKey FEATURE_GIANTS_ENABLED = new PluginConfigKey("features.giants.enabled", true);
    public static final PluginConfigKey FEATURE_GIANTS_BREAK_BLOCKS = new PluginConfigKey("features.giants.break-blocks", Arrays.asList(Material.GRASS.name(), Material.LEAVES.name(), Material.WOOD.name(), Material.GLASS.name(), Material.CROPS.name(), Material.SOIL.name(), Material.LOG.name(), Material.WOOD_STEP.name(), Material.WOOD_STAIRS.name()));

    public static final PluginConfigKey INTEGRATION_FACTIONS_PREVENT_SPAWING = new PluginConfigKey("features.integration.factions.prevent-spawing", true);
}
